### DISCUSSION TITLE
I implemented a proxy and Express middleware for Node.js

### DISCUSSION THREAD

**Discussion by marella** [grpc/grpc-web#1329] [Show and tell]
# [@grpc-web/proxy](https://github.com/marella/node-grpc-web/tree/main/packages/proxy) [![npm](https://img.shields.io/npm/v/@grpc-web/proxy)](https://www.npmjs.com/package/@grpc-web/proxy) [![tests](https://github.com/marella/node-grpc-web/actions/workflows/tests.yml/badge.svg)](https://github.com/marella/node-grpc-web/actions/workflows/tests.yml) [![coverage](https://coveralls.io/repos/github/marella/node-grpc-web/badge.svg)](https://coveralls.io/github/marella/node-grpc-web)

This is a standalone proxy server created using `@grpc-web/middleware` (see below), `connect` and `cors` packages. It passes the [interop tests](https://github.com/grpc/grpc-node/blob/master/test/interop/interop_client.js) and can be run directly from the command line:

```sh
npx @grpc-web/proxy --target http://localhost:9090 --listen 8080
```

or used as a library:

```js
const proxy = require('@grpc-web/proxy');

proxy({ target: 'http://localhost:9090' }).listen(8080);
```

It can be run in the same process as the gRPC server:

```js
// Proxy server
proxy({ target: 'http://localhost:9090' }).listen(8080);

// gRPC server
const server = new grpc.Server();
server.addProtoService(service, implementation);
server.bind('0.0.0.0:9090', credentials);
server.start();
```

For more details, please see [README](https://github.com/marella/node-grpc-web/tree/main/packages/proxy#readme).

# [@grpc-web/middleware](https://github.com/marella/node-grpc-web/tree/main/packages/middleware) [![npm](https://img.shields.io/npm/v/@grpc-web/middleware)](https://www.npmjs.com/package/@grpc-web/middleware) [![tests](https://github.com/marella/node-grpc-web/actions/workflows/tests.yml/badge.svg)](https://github.com/marella/node-grpc-web/actions/workflows/tests.yml) [![coverage](https://coveralls.io/repos/github/marella/node-grpc-web/badge.svg)](https://coveralls.io/github/marella/node-grpc-web)

This middleware can be used with existing Express and Connect applications. Here is a full working example of a proxy server created using `express` and `cors` packages:

```js
const express = require('express');
const cors = require('cors');
const grpcWeb = require('@grpc-web/middleware');

const app = express();
app.use(cors({ exposedHeaders: ['grpc-status', 'grpc-message'] }));
app.use(grpcWeb('http://localhost:9090'));
app.listen(8080);
```

It only handles gRPC Web requests and ignores normal HTTP requests, so it can be used together with normal web applications:

```js
app.use(grpcWeb('http://localhost:9090'));

app.get('/', (req, res) => {
  res.send('Hello World');
});
```

For more details, please see [README](https://github.com/marella/node-grpc-web/tree/main/packages/middleware#readme).

**Comment by sampajano** [comment]
Very cool.. such a simple solution.. Thanks very much for sharing..  :)

**Comment by fawzibazari** [comment]
Hello,

is it possible to use the package for server-side-streaming in a browser ?

can i use the  `@grpc-web/middleware` to replace envoy proxy ?

i did saw the http2 option, it will send http2 requests through express ?

Thanks

**Reply by fawzibazari** [reply]
Hello,
I forgot to write sooner i did use the package and it replace the Envoy server part but we have to use the grpc-web package and call the express proxy instead of the gRPC server 

Thanks !

**Reply by sampajano** [reply]
Thanks for sharing your experience! 😃
